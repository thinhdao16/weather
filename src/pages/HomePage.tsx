import axios from "axios"
import { useEffect, useMemo, useState } from "react"
import useBookStore from "../../store/bookStore"
import Search from "../components/Search"
function HomePage() {
    const [data, setData] = useState<any>()
    const [dataDay, setDataDay] = useState<any>()
    const { country } = useBookStore()
    const currrentDate = new Date()
    const options: any = { year: "numeric", month: "long", day: "numeric" }
    const formatDateToForm = (data: any) => {
        const format = data?.toLocaleDateString('en-US', options)
        return format
    }
    function formatDate(dateString: string) {
        const months = [
            "January", "February", "March", "April", "May", "June",
            "July", "August", "September", "October", "November", "December"
        ];
        const date = new Date(dateString);
        const year = date.getFullYear();
        const month = months[date.getMonth()];
        const day = date.getDate();
        return `${month} ${day}, ${year}`;
    }
    const apiKey = '85ca72f1b6e7246de11846015225c31d'


    const groupedSchedule = useMemo(() => {
        if (!dataDay || !dataDay.list) {
            return {};
        }
        return dataDay.list.reduce((acc:any, item:any) => {
            const day = item.dt_txt.split(' ')[0];
            const time = item.dt_txt.split(' ')[1].slice(0, -3);
            if (!acc[day]) {
                acc[day] = [];
            }
            acc[day].push({ ...item, time  });
            return acc;
        }, {});
     }, [dataDay]);
    useEffect(() => {
        const dataWeather = async () => {
            try {
                const [currentWeatherResponse, forecastResponse] = await Promise.allSettled([
                    axios.get(`https://api.openweathermap.org/data/2.5/weather?q=${country}&appid=${apiKey}`),
                    axios.get(`https://api.openweathermap.org/data/2.5/forecast?q=${country}&appid=${apiKey}`)
                ]);
                if (currentWeatherResponse.status === 'fulfilled') {
                    setData(currentWeatherResponse.value.data)
                } else {
                    console.error('Error fetching current weather:', currentWeatherResponse.reason);
                }
                if (forecastResponse.status === 'fulfilled') {
                    setDataDay(forecastResponse.value.data)
                } else {
                    console.error('Error fetching weather forecast:', forecastResponse.reason);
                }
            } catch (error) {
                console.error("get fail");
            }
        }
        dataWeather()
    }, [])


    return (
        <>
        <Search />
        <div className="bg-gray-200 px-6 py-4 flex flex-col justify-center items-center">           
            <div className=' w-full lg:w-1/4 md:w-1/2 sm:w-1/2'>
                <div className="p-4 rounded-xl bg-white shadow-md ">
                    <div className="text-start">
                        <span className="text-gray-500 font-medium">
                            {formatDateToForm(currrentDate)}
                        </span>
                    </div>
                    <div className="flex items-center justify-evenly">
                        <img src={`https://openweathermap.org/img/wn/${data?.weather[0]?.icon}@2x.png`} alt="error" />
                        <div className="flex flex-col">
                            <span className="text-4xl font-semibold" >{`${Math.floor(data?.main?.temp - 273)}°C`}</span>
                            <span className="text-gray-500 font-medium">
                                {`  ${data?.weather[0].description}`}
                            </span>
                        </div>

                    </div>
                    <div className="flex justify-between">
                        <div className="flex-1 text-gray-400">Humidity <span className="block text-black font-medium">
                            {data?.main?.humidity}%</span></div>
                        <div className="flex-1 text-gray-400">Winds <span className="block text-black font-medium">
                            {data?.wind?.speed} m/s</span></div>
                        <div className="flex-1 text-gray-400">Visibility <span className="block text-black font-medium">
                            {data?.visibility / 1000} km  </span></div>
                    </div>
                </div>
                <div className="text-start mt-4 mb-2">
                    <span className="font-medium">5-day Forecast (3 Hours)</span>
                </div>
                <div className="p-4 rounded-xl shadow-md bg-white flex flex-col text-start">
                    <div>
                        {Object.keys(groupedSchedule).map((day: any, index) => {
                            //    console.log(day)
                            return (
                                <div key={index}>
                                    <div className="text-gray-400">
                                        {(formatDateToForm(currrentDate) === formatDate(day)) ? (<div>Today</div>)
                                            :
                                            formatDate(day) 

                                        }
                                    </div>
                                    <div className="flex-col flex">
                                        {groupedSchedule[day].map((data: any, idx: any) => {
                                            return (

                                                <div className="grid grid-cols-12 items-center">
                                                    <span  key={idx} className="font-medium col-span-1" >{data?.time}</span>
                                                    <img src={`https://openweathermap.org/img/wn/${data?.weather[0]?.icon}@2x.png`} alt="" className="col-span-2" />
                                                    <span className="text-gray-400 col-span-4">${Math.floor(data?.main?.temp_max - 273)} / ${Math.floor(data?.main?.temp_min - 273)}°C</span>
                                               <div className="font-medium col-span-5 text-end">
                                               <span >{data?.weather[0]?.description}</span>

                                               </div>
                                                </div>
                                            )
                                        })}
                                    </div>
                                </div>
                            )
                        })}

                    </div>
                </div>
            </div>

        </div>
        </>
        


    )
}


export default HomePage
