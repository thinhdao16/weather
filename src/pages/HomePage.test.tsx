import HomePage from "./HomePage"
import { renderWithRouter } from "../test/utils/utils"



describe('<HomePage/>',()=>{
    test('render correctly',()=>{
        const {asFragment} = renderWithRouter(<HomePage/>)
        expect(asFragment()).toMatchSnapshot()
    })
})