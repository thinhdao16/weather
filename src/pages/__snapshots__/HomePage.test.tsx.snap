// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<HomePage/> > render correctly 1`] = `
<DocumentFragment>
  <div
    class="flex items-center justify-center "
  >
    <div
      class="w-full lg:w-1/4 md:w-1/2 sm:w-1/2 px-6 py-2 flex"
    >
      <button
        class="flex-1 text-left flex gap-1 items-center"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g
            id="Location_On"
          >
            <g>
              <path
                d="M12,21.933a1.715,1.715,0,0,1-1.384-.691L5.555,14.5a7.894,7.894,0,1,1,12.885-.009L13.385,21.24A1.717,1.717,0,0,1,12,21.933ZM11.992,3.066A6.81,6.81,0,0,0,7.414,4.815a6.891,6.891,0,0,0-1.05,9.1l5.051,6.727a.725.725,0,0,0,.584.292h0a.732.732,0,0,0,.586-.292l5.044-6.734A6.874,6.874,0,0,0,12.81,3.113,7.277,7.277,0,0,0,11.992,3.066Z"
              />
              <path
                d="M12,12.5A2.5,2.5,0,1,1,14.5,10,2.5,2.5,0,0,1,12,12.5Zm0-4A1.5,1.5,0,1,0,13.5,10,1.5,1.5,0,0,0,12,8.5Z"
              />
            </g>
          </g>
        </svg>
        <span
          class="font-semibold"
        >
          Ho Chi Minh City, VN
        </span>
      </button>
      <div
        class="flex-1 flex items-center justify-end text-right"
      >
        <button>
          <svg
            fill="currentColor"
            height="1em"
            id="search"
            stroke="currentColor"
            stroke-width="0"
            version="1.1"
            viewBox="0 0 24 24"
            width="1em"
            x="0px"
            xmlns="http://www.w3.org/2000/svg"
            y="0px"
          >
            <g>
              <path
                d="M20.031,20.79c0.46,0.46,1.17-0.25,0.71-0.7l-3.75-3.76c1.27-1.41,2.04-3.27,2.04-5.31
		c0-4.39-3.57-7.96-7.96-7.96s-7.96,3.57-7.96,7.96c0,4.39,3.57,7.96,7.96,7.96c1.98,0,3.81-0.73,5.21-1.94L20.031,20.79z
		 M4.11,11.02c0-3.84,3.13-6.96,6.96-6.96c3.84,0,6.96,3.12,6.96,6.96c0,3.84-3.12,6.96-6.96,6.96C7.24,17.98,4.11,14.86,4.11,11.02
		z"
              />
            </g>
          </svg>
        </button>
      </div>
    </div>
  </div>
  <div
    class="bg-gray-200 px-6 py-4 flex flex-col justify-center items-center"
  >
    <div
      class=" w-full lg:w-1/4 md:w-1/2 sm:w-1/2"
    >
      <div
        class="p-4 rounded-xl bg-white shadow-md "
      >
        <div
          class="text-start"
        >
          <span
            class="text-gray-500 font-medium"
          >
            July 26, 2024
          </span>
        </div>
        <div
          class="flex items-center justify-evenly"
        >
          <img
            alt="error"
            src="https://openweathermap.org/img/wn/undefined@2x.png"
          />
          <div
            class="flex flex-col"
          >
            <span
              class="text-4xl font-semibold"
            >
              NaNÂ°C
            </span>
            <span
              class="text-gray-500 font-medium"
            >
                undefined
            </span>
          </div>
        </div>
        <div
          class="flex justify-between"
        >
          <div
            class="flex-1 text-gray-400"
          >
            Humidity 
            <span
              class="block text-black font-medium"
            >
              %
            </span>
          </div>
          <div
            class="flex-1 text-gray-400"
          >
            Winds 
            <span
              class="block text-black font-medium"
            >
               m/s
            </span>
          </div>
          <div
            class="flex-1 text-gray-400"
          >
            Visibility 
            <span
              class="block text-black font-medium"
            >
              NaN km  
            </span>
          </div>
        </div>
      </div>
      <div
        class="text-start mt-4 mb-2"
      >
        <span
          class="font-medium"
        >
          5-day Forecast (3 Hours)
        </span>
      </div>
      <div
        class="p-4 rounded-xl shadow-md bg-white flex flex-col text-start"
      >
        <div />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
